versioned_docs/version-2020fa/lecture8.md

```diff
@@ -3,4 +3,561 @@ id: lecture8
+
+Axios is an npm package that can give you a better experience fetching data.
+
+It has methods like:
+
+- `axios.get(url[, config])`
+- `axios.post(url[, data[, config]])`
+- `axios.put(url[, data[, config]])`
+- `axios.delete(url[, config])`
+
+Axios also returns a promise and can be resolved with `.then()`.
+
+Here's an example:
+
+```typescript
+// fetch()
+const fetchData = async () => {
+  const response = await fetch('https://jsonplaceholder.typicode.com/posts');
+  const posts = await response.json();
+  setData(posts);
+};
+
+// axios
+const fetchData = async () => {
+  const posts = await axios.get('https://jsonplaceholder.typicode.com/posts');
+  setData(posts);
+};
+```
+
+In Typescript you can provide type definitions for the response you'll get from
+a request, similar to what we do in `useState<type>`. If you don't specify,
+you'll get the `any` type.
+
+```typescript
+// anyResponse: AxiosResponse<any>
+const anyResponse = await axios.get('url');
+// anyData: any
+const anyData = anyResponse.data;
+
+type DataType = ...;
+// typedResponse: AxiosResponse<DataType>
+const typedResponse = await axios.get<DataType>('url');
+// typedData: DataType
+const typedData = typedResponse.data;
+```
+
+## CORS workaround: proxy
+
+When you are testing your backend and frontend together on localhost, you will
+come across some sort of **CORS policy error** when you try to call the backend
+from the frontend.
+
+One way to get around this is to add this line to your **frontend**'s
+`package.json`:
+
+```json
+  "proxy": "http://localhost:8080",
+```
+
+where `8080` is the the port of your **backend** (change it accordingly).
+
+After adding the proxy config, remove the base URL from your requests:
+
+```typescript
+// do this
+const posts = await axios.get('/getAllPosts');
+
+// instead of this
+const posts = await axios.get('http://localhost:8080/getAllPosts');
+```
+
+## Summary
+
+You keep track of your data in component state. Your app displays whatever data
+you have.
+
+You update your data by calling an endpoint within `useEffect` and setting your
+data to the response that you get back.
+
+You can call endpoints using `fetch()` or `axios` and handle the responses
+asynchronously.
+
+## Demo Code
+
+### Backend
+
+```typescript title="index.ts (backend)"
+import admin from 'firebase-admin';
+import express from 'express';
+
+const serviceAccount = require('./service-account.json');
+
+admin.initializeApp({
+  credential: admin.credential.cert(serviceAccount),
+  databaseURL: '[DATABASE_URL]',
+});
+
+const db = admin.firestore();
+
+const app = express();
+const port = 8080;
+app.use(express.json());
+
+type Song = {
+  name: string;
+  artist: string;
+  rating: number;
+};
+
+type SongWithID = Song & {
+  id: string;
+};
+
+const songsCollection = db.collection('songs');
+
+app.get('/getSongs', async (req, res) => {
+  const songs = await songsCollection.get();
+  res.json(
+    songs.docs.map(
+      (doc): SongWithID => {
```

Comment:
this type annotation is wrong. you should just remove it

Reply:
It's typing the return value of the function, not `doc`. It looks right to me, but I might be tripping.

Reply:
oops you are right